pkg/migrator/migrator.go

```diff
@@ -39,115 +39,141 @@ func NewMigrator(client client.Client) migrator {
 	return migrator{
 		logger: log.NewEntry(log.New()),
 		client: client,
+		retry:  3,
 	}
 }
 
 type migrator struct {
 	logger *logrus.Entry
 	client client.Client
+	retry  int
 }
 
-// Migrate updates existing Subscriptions, deletes the CSCs installed during
+// doMigrate updates existing Subscriptions, deletes the CSCs installed during
 // operator installation as well the stale datastore CSCs when the cluster is
-// migrating from Openshift 4.1.x to Openshift 4.2.x
+// migrating from Openshift 4.1.z to Openshift 4.2.z
+func (m *migrator) doMigrate(operatorNamespace string) []string {
+	installedCscs, errorList := m.updateSubscriptions()
+	if errorList != nil {
+		return errorList
+	}
+	errorList = m.deleteInstalledCscs(installedCscs, operatorNamespace)
+	if errorList != nil {
+		return errorList
+	}
+	errorList = m.deleteDatastoreCscs(operatorNamespace)
+	return errorList
+}
+
+// Migrate is a wrapper for doMigrate so that we can retry the migration in case
+// of failures.
 func (m *migrator) Migrate(operatorNamespace string) {
-	installedCscs := m.updateSubscriptions()
-	m.deleteInstalledCscs(installedCscs, operatorNamespace)
-	m.deleteDatastoreCscs(operatorNamespace)
+	errorList := m.doMigrate(operatorNamespace)
+	if errorList != nil {
+		for _, err := range errorList {
+			m.logger.Errorf(err)
+		}
+		m.logger.Infof("[migration] Failed to migrate")
+		m.retry = m.retry - 1
+		if m.retry > 0 {
```

Comment:
Won't a `for` loop be more cleaner? I don't see a reason to do a recursion.

Reply:
Yeah there was no specific reason for me to do this recursively - just habit